arch/xtensa: Add a call0 ABI variant

andyross opened the pull request:
New context layer for building Zephyr with the call0 ABI.  The system call overhead of having to spill 64 GPRs, and the general complexity of the existing asm2 layer, is prohibitive when trying to build a userspace kernel entry path for Xtensa.

Note: it still uses the windowed registers, but as a quick-switch register set for optimized interrupt entry.  This is not a mechanism for building Zephyr on hardware that actually lacks windows (though with some work it could be).

This is much (much) leaner in terms of entry/exit overhead, is significantly smaller, and should scale better with changes needed for userspace.  There is a cost in terms of function call code size and execution speed, but in practice it seems manageable (I'm measuring ~4% with a no-frame-pointer build, Max Filippov reports more like ~13% in Linux builds -- both numbers are well under the kind of overhead we see enabling userspace at all).

andyross committed:
arch/xtensa: Rename "ALLOCA" ZSR to "A0SAVE"

This register alias was originally introduced to allow A0 to be used
as a scratch register when handling exceptions from MOVSP
instructions. (It replaced some upstream code from Cadence that
hard-coded EXCSAVE1).  Now the MMU code is now using too, and for
exactly the same purpose.

Calling it "ALLOCA" is only confusing.  Rename it to make it clear
what it's doing.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
drivers/console: xtensa_sim_console: implement arch_printk_char_out()

This is an older driver and didn't support the weak
arch_printk_char_out() hook, which is a link-time symbol that allows
logging to work from the first instruction.  Some drivers can't do
that because they need an initialization step, but this one works
great.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Add XTENSA_CALL0_ABI kconfig

New ABI, much simpler, slightly bigger/slower code, but MUCH cleaner
syscall entry.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: #include cleanup

This file doesn't need the asm2 header, and the preprocessor logic
around whether to include the backtrace header is needless (all it
does is declare functions).

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Add a call0 ABI variant

New context layer for building Zephyr with the call0 ABI.  The system
call overhead of having to spill 64 GPRs, and the general complexity
of the existing asm2 layer, is prohibitive when trying to build a
userspace kernel entry path for Xtensa.

Note: it still uses the windowed registers, but as a quick-switch
register set for optimized interrupt entry.  This is not a mechanism
for building Zephyr on hardware that actually lacks windows (though
with some work it could be).

This is much (much) leaner in terms of entry/exit overhead, is
significantly smaller, and should scale better with changes needed for
userspace.  There is a cost in terms of function call code size and
execution speed, but in practice it seems manageable (I'm measuring
~4% with a no-frame-pointer build, Max Filippov reports more like ~13%
in Linux builds -- both numbers are well under the kind of overhead we
see enabling userspace at all).

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
hello_world rig for exercising xtensa-win0

andyross commented:
See docs in the PR itself, but I'll paste here for visibility:

# Zephyr with the Xtensa CALL0 ABI

The Xtensa register window mechanism is a poor fit for memory
protection.  The existing Zephyr optimizations, in particular the
cross-stack call we do on interrupt entry, can't be used.  The way
Xtensa windows spill is to write register state through the stack
pointers it finds in the hidden caller registers, which are wholely
under the control of the userspace app that was interrupted.  The
kernel can't be allowed to do that, it's a huge security hole.

Naively, the kernel would have to write out the entire 64-entry
register set on every entry from a context with a PS.RING value other
than zero, including system calls.  That's not really an acceptable
performance or complexity cost.

Instead, for userspace apps, Zephyr builds using the "call0" ABI,
where only a fixed set of 16 GPRs (see section 10 of the Cadence
Xtensa Instruction Set Architecture Summary for details on the ABI).
Kernel traps can then use the remaining GPRs for their own purposes,
greatly speeding up entry speed.

## Toolchain

Existing Xtensa toolchains support a `-mabi=call0` flag to generate
code with this ABI, and it works as expected.  It sets a
__XTENSA_CALL0_ABI__ preprocessor flag and the Xtensa HAL code and our
crt1.S file are set up to honor it appropriately (though there's a
glitch in the Zephyr HAL integration, see below).

Unfortunately that doesn't extend to binary artifacts.  In particular
the libgcc.a files generated for our existing SDK toolchains are using
the windowed ABI and will fail at runtime when you hit 64 bit math.

Cadence toolchains have an automatic multilib scheme and will select a
compatible libgcc automatically.

But for now, you have to have a separate toolchain (or at least a
separately-built libgcc) for building call0 apps.  I'm using this
script and it works fine, pending proper SDK integration:

```shell
#!/bin/sh
set -ex

TC=$1
if [ -z "$TC" ]; then
TC=sample_controller
fi

# Grab source (these are small)
git clone https://github.com/zephyrproject-rtos/sdk-ng
git clone https://github.com/crosstool-ng/crosstool-ng

# Build ct-ng itself
cd crosstool-ng
./bootstrap
./configure --enable-local
make -j$(nproc)

# Configure for the desired toolchain
ln -s ../sdk-ng/overlays
cp ../sdk-ng/configs/xtensa-${TC}_zephyr-elf.config .config

grep -v CT_TARGET_CFLAGS .config > asdf
mv asdf .config
echo CT_TARGET_CFLAGS="-mabi=call0" >> .config
echo CT_LIBC_NONE=y >> .config

./ct-ng olddefconfig
./ct-ng build.$(nproc)

echo "##"
echo "## Set these values to enable this toolchain:"
echo "##"
echo export CROSS_COMPILE=$HOME/x-tools/xtensa-${TC}_zephyr-elf/bin/xtensa-${TC}_zephyr-elf-
echo export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
```

Note you don't really need to use the toolchain that was built, it's
enough to take the libgcc.a and drop it on top of the one in your SDK,
just be careful to save the original, etc...

Longer term we need proper integration into the SDK, obviously.
Ideally we would just rebuild libgcc in isolation, but I couldn't
figure out how to do that except in the context of a full gcc build.
We could do the separate build in sdk-ng and then copy the files to a
multilib scheme of our own devising, which is something existing
Zephyr platforms sometimes need to do.  Also I note that other gcc
architectures have already solved problems like this (c.f. x86_64 and
x32 working in the same toolchain, etc...), but couldn't find any
Xtensa-specific support.

Also as a quick start: it's possible to build Zephyr in such a way as
to avoid (almost/maybe) all calls to libgcc via disabling kconfigs
that trigger 64 bit code.  I found I could get full coverage of the
new call0 code with a custom test rig and `-- -DCONFIG_TIMEOUT_64BIT=n -DCONFIG_CBPRINTF_REDUCED_INTEGRAL=y -DCONFIG_CBPRINTF_NANO=y`

## Bugs and Missing Features

No support in the SDK.  See above about toolchains.

The existing syscall implementation is a mock, because this happened
in parallel with the new MMU-enabled dc233c platform.  The only code
difference is that it uses a single static region loaded from a
"_mock_priv_stack" symbol instead of a real thread kernel stack.  In
theory everything will work when you enable CONFIG_USERSPACE, but...

Likewise the MMU integration is a blindly-copied stub and unexercised.
It's really simple though ("just read from PTEVADDR with PS.EXCM=1 to
prime the TLB"), so hopefully it won't need much fixing.

There's no C implementation of the Zephyr "call syscalls" API, just
assembly.  The interface is deliberately based on the call0 function
call ABI though, so this should be trivially easy.

For simplicity, this code is written to save context to the arch
region of the thread struct and not the stack (except for nested
interrupts, obviously).  That's a common pattern in Zephyr, but for
KERNEL_COHERENCE (SMP) platforms it's actually a performance headache,
because the stack is cached where the thread struct is not.  We should
move this back to the stack, but that requires doing some logic in the
assembly to check that the resulting stack pointer doesn't overflow
the protected region, which is slightly non-trivial (or at least needs
a little inspiration).

Right now the Zephyr HAL integration doesn't build when the call0 flag
is set because of a duplicated symbol.  I just disabled the build at
the cmake level for now, but this needs to be figured out.

The ARCH_EXCEPT() handling is a stub and doesn't actually do anything.
Really this isn't any different than the existing code, it just lives
in the asm2 files that were disabled and I have to find a shared
location for it.

Backtrace logging is likewise specific to the older frame format and
ABI and doesn't work.  The call0 ABI is actually much simpler, though,
so this shouldn't be hard to make work.

FPU support isn't enabled.  Likewise this isn't any different (except
trivially -- the context struct has a different layout).  We just need
to copy code and find compatible hardware to test it on.

A good security audit is needed to make sure all the holes are
plugged.  At the very least, note that when the PS.WOE bit is set,
user thread contexts are able to "rotate" registers and write to their
contexts by doing regular function calls.  We need to be 100% sure no
RING 1+ code ever runs with this set.  (Also the ISA spec is a little
ambiguous as to how the hardware treats CALLn/ENTRY/RETW instructions
when WOE=0, I worry a tiny bit there might be hardware that has
unpluggable holes...)

I realized when writing this that our existing handling for NMI
exceptions (strictly any level > EXCM_LEVEL) isn't correct, as we
generate ZSR-style EPS/EPC register usage in those handlers that isn't
strictly separated from the code it may/will have interrupted.  This
is actually a bug with asm2 also, but it's going to make
high-priority/DEBUG/NMI exceptions unreliable unless fixed (I don't
think there are any Zephyr users currently though).

andyross commented:
The context follows discussion from #61303 so be sure to check there also.

Basically: it's real and it works.  The code is comparatively tiny vs. the maybe-too-clever-for-its-own-good asm2 integration (which I still really like, but really isn't a good fit for privilege separation).  I'm very happy on the whole.  The libgcc issue is going to be a pain, and I'm really hoping someone with more SDK fu and toolchain interest can help with that?   @jcmvbkbc 's name is all over the GNU toolchain patches for the architecture, so he surely has suggestions on how to make this work best.

The only missing feature still causing any test failures I can find is ARCH_EXCEPT(), which I should have patched up RSN (it's just a code motion thing).  The other missing stuff is all in the nice-to-have category and IMHO shouldn't block merge or further work.

This was all done against sample_controller, with a few tries for validation on my mt8195 at various stages.  It should ("should") work everywhere but in practice there will be burps.   First up on my list is to get it working on dc233c to validate the MMU handling, though that will need to wait a few days while I catch up on my day job tasks which have been woefully neglected while I work on fun stuff.

andyross commented:
(added a selection of Xtensa stakeholders, please pull in anyone I've forgotten)

Also it's worth pointing out that at least a little thought has been given to a re-unification path, since I don't think people will want to give up on windows for platforms where they provide value, and no one wants to maintain two completely separate context layers in the long term.

Most of the code isn't involved in the "register-twirl on interrupt" trick at all.  We can partition things out into paths like "pre-call context save" (which can be e.g. save-all-64, or twirl, etc...), "do the call" (which can be a simple ABI call or the cross stack trick), "post-call-switch", "post-call-no-switch", etc...  The state needed to discriminate that is really just the RING bits of PS, so the overhead isn't awful even if getting it to generate optimally might be a pain.  Definitely not for the first version but maybe at some point.

andyross commented on arch/xtensa/core/xtensa-win0.S:
```diff
@@ -0,0 +1,481 @@
+	 */
+	movi a1, ~(PS_RING_MASK | PS_UM_MASK | PS_INTLEVEL_MASK)
+	and a0, a0, a1
+	addi a0, a0, 1 /* INTLEVEL = 1 */
```
I'm realizing while re-reading this post-submit that this is actually wrong.  I blindly copied the need to adjust intlevel from the L1 handler, but in fact syscalls (after setup) are **thread** contexts and should be running at an interrupt level of zero, obviously.  My mock rig didn't catch that, but obviously the test suite will start flagging errors if you're not able to receive L1 interrupts from syscalls.

Should work to just remove that one line and adjust all the comments talking about this.

dcpleung commented:
FYI, I did a quick tinkering with sdk-ng trying to build with multilib. However, the usually way of specifying `MULTILIB_OPTIONS` was not successful. More investigation is needed to see if it is even possible to have multilib support.

andyross commented:
> FYI, I did a quick tinkering with sdk-ng trying to build with multilib.

You got farther than I did.  I discovered those MULTILIB_ argument hooks via google, saw that they weren't touched by the xtensa stuff, and gave up hoping quite frankly that you'd know how to fix it. :) But we also have Max's attention here who's (I assume) the maintainer, so maybe there's an even better resource. @jcmvbkbc ?

jcmvbkbc commented:
> maybe there's an even better resource

Not at the moment, unfortunately. In linux it was solved by getting rid of the libgcc dependency and adding all required helper sources to the kernel.

andyross commented:
> In linux it was solved by getting rid of the libgcc dependency and adding all required helper sources to the kernel.

Oh hey, well look at that, there they are.  That's really clever.  Unfortunately we're likely to run into licensing issues with importing GPL code in Zephyr (but I'm **really** not the expert on this stuff); any idea if the source is available in Apache-compatible form anywhere?  But the code in the Linux tree is a TON simpler (and smaller -- looks like most of libgcc is never used by xtensa compiler output?) than the original libgcc, so if nothing else it would be a good guide for us building our own artifacts for the SDK without building a whole compiler to throw away.

jcmvbkbc commented:
> any idea if the source is available in Apache-compatible form anywhere?

I only know of GPL-licensed version of this code. OTOH LLVM should have something equivalent, perhaps under different license.

dcpleung commented:
Hm... I am trying to run `hello_world` with both sample_controller and I got:

```
*** Booting Zephyr OS build zephyr-v3.4.0-2857-g5a5738736355 ***
Hello World! qemu_xtensa
In main thread  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU 
```

Do I need some other kconfig other than `CONFIG_XTENSA_CALL0_ABI=y`?

dcpleung commented:
Note: you actually need a toolchain built with CALL0 to work, or else the mock-up `hello_world` would crash with unrecoverable CPU exceptions.

andyross commented:
> Note: you actually need a toolchain built with CALL0 to work, or else the mock-up hello_world would crash with unrecoverable CPU exceptions.

That's exactly the errors you get with the windowed libgcc (it panics on the ENTRY instruction at the first 64 bit operation, then you get a recursive panic trying to log it IIRC).

I swear I had that working by dropping the libgcc.a from the rebuilt toolchain on top of the SDK version though.  But indeed it's probably simplest to use a rebuilt one.

andyross commented:
Also thinking on this a bit: we're actually fine with licenses.  We can't put GPL code into the Zephyr tree itself, but we have lots of modules with GPL code in them already.  Presumably the existing HAL module would be a good spot.  And even in the worst case, we can put it in the SDK.  The SDK is **already** distrubuting (built) GPLv3 code by definition.  In point of fact we're actually required by the license to make the source code available to anyone who asks.

The only question is which version (Linux or gcc) of the code we pull and how much work it will take to make it portable.  I did a quick check:

+ The xtensa assembly for GNU libgcc is really clean, lives in just two files, and parametrized only by core-isa.h symbols (though the header is actually named something else).  It builds very cleanly with minimal change.

+ The Linux assembly is less pleasing, honestly.   It's split up into a bunch of files and has a bunch of Linuxisms embedded.

+ The remaining symbols we're missing (by comparison with the smaller Linux libgcc library) are in C though, and that source is less portable (in both GNU and Linux)  Lots of special type names and #if'ery would have to be reverse engineered.   I could only find three undefined symbols in there, so maybe this wouldn't be too much work.

Anyway I think Linux has this right: we should import and build the source to libgcc ourselves.  It's for sure easier than dealing with toolchain headaches.

FWIW: I did look at LLVM too.  They have a "compiler_rt" library to do what libgcc does, and AFAICT the symbol names are mostly identical.  But lots of architectures (including xtensa) are missing asm hooks and only build the C versions of the routines.  I think what's happening actually is that for the shorter utilities LLVM is actually injecting IR directly into the code stream and not linking against anything at all.  Which is really clever but completely useless to us trying to build with gcc.

dcpleung commented:
Got GCC multilib to work: https://github.com/zephyrproject-rtos/sdk-ng/pull/701 and https://github.com/zephyrproject-rtos/gcc/pull/23. I have tried it locally and they seem to work correctly.

andyross commented:
Me:
> we should import and build the source to libgcc ourselves. It's for sure easier than dealing with toolchain headaches.

[12 minutes later]

Daniel:
> Got GCC multilib to work [in two patches and like six lines of code]

... and this is why you don't trust me to pontificate about toolchains, I guess.

dcpleung commented:
> Me:
> 
> > we should import and build the source to libgcc ourselves. It's for sure easier than dealing with toolchain headaches.
> 
> [12 minutes later]
> 
> Daniel:
> 
> > Got GCC multilib to work [in two patches and like six lines of code]
> 
> ... and this is why you don't trust me to pontificate about toolchains, I guess.

and 2 days of cursing...

dcpleung commented:
Ran latency measurements and this seems (mostly) a bit faster:

Windowed:
```
Timing results: Clock frequency: 1 MHz
Average thread context switch using yield           :     289 cycles ,   289420 ns : 
Average context switch time between threads (coop)  :     297 cycles ,   297571 ns : 
Switch from ISR back to interrupted thread          :     108 cycles ,   108000 ns : 
Time from ISR to executing a different thread       :     166 cycles ,   166000 ns : 
Time to create a thread (without start)             :     304 cycles ,   304000 ns : 
Time to start a thread                              :     311 cycles ,   311000 ns : 
Time to suspend a thread                            :     353 cycles ,   353000 ns : 
Time to resume a thread                             :     334 cycles ,   334000 ns : 
Time to abort a thread (not running)                :      37 cycles ,    37000 ns : 
Average semaphore signal time                       :      45 cycles ,    45445 ns : 
Average semaphore test time                         :      17 cycles ,    17285 ns : 
Semaphore take time (context switch)                :     394 cycles ,   394000 ns : 
Semaphore give time (context switch)                :     421 cycles ,   421000 ns : 
Average time to lock a mutex                        :      23 cycles ,    23045 ns : 
Average time to unlock a mutex                      :      12 cycles ,    12850 ns : 
Average time for heap malloc                        :     387 cycles ,   387930 ns : 
Average time for heap free                          :     357 cycles ,   357800 ns : 
```

CALL0:
```
Timing results: Clock frequency: 1 MHz
Average thread context switch using yield           :     203 cycles ,   203622 ns : 
Average context switch time between threads (coop)  :     212 cycles ,   212459 ns : 
Switch from ISR back to interrupted thread          :      55 cycles ,    55000 ns : 
Time from ISR to executing a different thread       :     118 cycles ,   118000 ns : 
Time to create a thread (without start)             :     305 cycles ,   305000 ns : 
Time to start a thread                              :     218 cycles ,   218000 ns : 
Time to suspend a thread                            :     236 cycles ,   236000 ns : 
Time to resume a thread                             :     236 cycles ,   236000 ns : 
Time to abort a thread (not running)                :      31 cycles ,    31000 ns : 
Average semaphore signal time                       :      41 cycles ,    41605 ns : 
Average semaphore test time                         :      24 cycles ,    24965 ns : 
Semaphore take time (context switch)                :     257 cycles ,   257000 ns : 
Semaphore give time (context switch)                :     300 cycles ,   300000 ns : 
Average time to lock a mutex                        :      30 cycles ,    30725 ns : 
Average time to unlock a mutex                      :      17 cycles ,    17321 ns : 
Average time for heap malloc                        :     322 cycles ,   322610 ns : 
Average time for heap free                          :     305 cycles ,   305920 ns : 
```

andyross commented:
Nice, that looks almost exactly the way I would have hoped it would: the big numbers (the ones that involve trips through new context code) are all faster by 10-30% because the new code is shorter since it deals with less state[1] (even cross stack calls need to do a bunch of bookeeping to make it work).

The small numbers are things like "lock a semaphore without context switch", and that's just measuring local code speed of tiny leaf functions.  And those are going to be necessarily slower, since leaf calls in call0 need to spill/fill all the caller-save registers and can't rely on the register windows to buffer for them.  (This is actually sort of worst case for call0.)

The one outlier I can see is that the heap tests got... measurably faster?  I don't have an answer there.  There's minimal kernel involvement in the heap (unless you have to wait on a full heap to fill an allocation, which clearly the test isn't).  That should be just plain old code and thus somewhat slower with call0.  No idea, but the delta is big enough that I'm suspicious we might have something wrong with the test.

[1] Also, to be blunt: it's tighter code because I did it better the second time around now that I know the architecture better.  It's not like I'm not proud of asm2, but it definitely shows the scars of early mistakes.

dcpleung commented:
Massaged enough code to get userspace hello_world to run: https://github.com/dcpleung/zephyr/tree/xtensa/call0_userspace. I mixed this with the userspace code from @ceolin as I was too lazy to cherry-pick the changes over.

andyross committed:
arch/xtensa: Rename xtensa_mmu.c to ptables.c

The job of this file is almost exclusively page table management.  MMU
hardware interaction is in there but a small part, and they don't
interact except in a few small APIs.  Rename it in advance of doing
work to the hardware side.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Add new MMU layer

This is a reworked MMU layer, sitting cleanly below the page table
handling in the OS, intended to integrate with call0 ABI handling and
to be more performant and efficient.  Notable differences from the
original work:

+ Significantly smaller code and simpler API (just three functions to
  be called from the OS/userspace/ptable layer).

+ Big README-MMU document containing my learnings over the process, so
  hopefully fewer people need to go through this in the future.

+ No TLB flushing needed.  Clean separation of ASIDs, just requires
  that the upper levels match the ASID to the L1 page table page
  consistently.

+ Validation layer to check page table caching attributes.  Caching
  the PTEs is poison to SMP, as it requires a data cache flush on
  every CPU for every change of a page table anywhere.  Even on
  single-core systems, having one mapping be write-through and the
  other writeback (which is what we had!) can break things.  Hopefully
  this catches all the mistakes.

+ Vector mapping is done with a 4k page and not a 4M page, leading to
  much more flexibility with hardware memory layout.  The original
  scheme required that the 4M region containing vecbase be mapped
  virtually to a location other than the hardware address, which makes
  confusing linkage with call0 and difficult initialization
  constraints where the exception vectors run at different addresses
  before and after MMU setup (effectively forcing them to be PIC
  code).

+ More provably correct initialization, all MMU changes happen in a
  single asm block with no memory accesses which would generate a
  refill.

+ Requires CONFIG_XTENSA_CALL0_ABI.  This isn't strictly an issue of
  behavior: the MMU code doesn't care about reguster use per se.  But
  with the register-windowed context management code in the older
  asm2, it's impossible to provide security across memory domains.
  Just rip the bandaid off.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
soc/xtensa/dc233c: Link iram/iram0 sections generically

The linkage for the "after vector" RAM region was specifying files
explicitly, which breaks when other files (or new files) want to add
code to these regions.  In particular call0 needs to use iram to store
local jump targets for the vector entry.

There's really no good spec on this.  Conventionally xtensa platforms
have always put the .iram and iram0 sections (and their derived
sections like literals) immediately after the vector table, but I
doubt there's a promise to that extent anywhere.

Just try to adhere to what convention there is.

(I note there's also a ".iram1" section referenced here that I'm not
moving because I don't know what's in it)

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Don't dump unsupported backtraces with call0

The backtrace mechanism is specific to the windowed call ABI
currently, so doesn't work with call0.  This was hidden previously
because backtraces were disabled by default at the kconfig level for
the old ("sample_controller") qemu_xtensa, but are on now with dc233c
for essentially arbitrary reasons.  There was some protection against
this already, extend it to the case where "we have windows, but aren't
using them".

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Add call0-based syscall interface

There was a system call handler added as part of the call0 ABI
integration.  Wire it up to the standard Zephyr userspace system call
API.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch/xtensa: Clean up MMU initialization

We were trying to initialize the MMU extremely early (out of crt1.S),
which was forcing some odd gymnastics with splitting up the
initialization paths.  Just don't do that.

The MMU boots with a pefectly good mapping for physical memory.  Use
that until we're set up and about to launch the post-arch
initialization in z_cstart().

Just in general, trying to run compiled C code before z_cstart() calls
kernel_arch_init() is asking for trouble.  We do it only in extreme
situations, and we don't need it here.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Another spin, now with full MMU support for dc233c.

This is intended to go under a rework of the userspace PR (which I
need to find a place to put).  As it is it splits the existing MMU
code into a mostly-unchanged "page table" layer and an entirely new
hardware driver for the bottom bits.  Paste major deltas from the
commit message:

+ Significantly smaller code and simpler API (just three functions to
  be called from the OS/userspace/ptable layer).

+ Big README-MMU document containing my learnings over the process, so
  hopefully fewer people need to go through this in the future.

+ No TLB flushing needed.  Clean separation of ASIDs, just requires
  that the upper levels match the ASID to the L1 page table page
  consistently.

+ Validation layer to check page table caching attributes.  Caching
  the PTEs is poison to SMP, as it requires a data cache flush on
  every CPU for every change of a page table anywhere.  Even on
  single-core systems, having one mapping be write-through and the
  other writeback (which is what we had!) can break things.  Hopefully
  this catches all the mistakes.

+ Vector mapping is done with a 4k page and not a 4M page, leading to
  much more flexibility with hardware memory layout.  The original
  scheme required that the 4M region containing vecbase be mapped
  virtually to a location other than the hardware address, which makes
  confusing linkage with call0 and difficult initialization
  constraints where the exception vectors run at different addresses
  before and after MMU setup (effectively forcing them to be PIC
  code).

+ More provably correct initialization, all MMU changes happen in a
  single asm block with no memory accesses which would generate a
  refill.

+ Requires CONFIG_XTENSA_CALL0_ABI.  This isn't strictly an issue of
  behavior: the MMU code doesn't care about reguster use per se.  But
  with the register-windowed context management code in the older
  asm2, it's impossible to provide security across memory domains.
  Just rip the bandaid off.

jcmvbkbc-issues commented on arch/xtensa/core/xtensa-win0.S:
```diff
@@ -0,0 +1,476 @@
+_DoubleExceptionVector:
+	wsr a0, ZSR_A0SAVE
+	rsr a0, PTEVADDR
+	l32i a0, a0, 0
```
a0 goes nowhere, is anything missing here?

andyross replied:
This is basically a mis-split.  There was a vestigial handler here in the first version of the PR that just blindly duplicated the existing in-tree handler, which itself was just doing the "load from PTEVADDR to prime the data cache with a page table page" trick (so indeed A0 gets dropped; the side effect on the TLB is what we want).  But that wasn't ever correct to begin with, and a later patch in this PR then replaces it with a jump into the L1 exception handler instead (so later userspace code can handle the fault just like any other memory fault).

There's basically no reason to have this here at all at this point in the PR, I'll remove in the next respin.

jcmvbkbc-issues commented on arch/xtensa/core/xtensa-win0.c:
```diff
@@ -0,0 +1,157 @@
+	irqs &= intenable;						\
+	while ((m = _xtensa_handle_one_int##l(irqs))) {			\
+		irqs ^= m;						\
+		__asm__ volatile("wsr.intclear %0" : : "r"(m));		\
```
This logic (first handle an interrupt / then clear IRQ bit) does not work very well with edge-triggered interrupts, because and IRQ may be set in the gap between the handler completion and wsr.intclear.

andyross replied:
That's a valid point; this is actually ancient code forklifted over from the original in asm2, and frankly I don't think any of us gave much thought to edge sensitivity.  Just moving the clear to before the handler can be racy with level triggering though if the device doesn't have its own mask to clear in the handler.   Maybe we could handle this with an assert just to disallow edge interrupts for now.

Really the right thing would be to detect the interrupt type at generation time and put the clear there.  We actually do have code to do this, though it's crufty and needs some rework of its own...

andyross replied:
I swear I have it in my head somewhere that INTCLEAR is used as a bus acknowledgement (i.e. visible to the device) for at least some classes of interrupts?  In which case you can get the converse problem: a device can see the ACK and trigger another interrupt before then handler had run, and the handler would see one event where the device signaled two.

But maybe that's wrong?  This is just vague memory on my part and you're the expert.  If writing INTCLEAR before the handler is safe in all cases then yeah, we should fix this.

andyross replied:
OK, maybe the best thing is to file this as a bug ("Edge triggered interrupts are racy on xtensa") and fix it in parallel.

ceolin commented on arch/xtensa/core/xtensa-win0.S:
```diff
@@ -0,0 +1,523 @@
+ *
+ * + System calls have their own optimized handler
+ * + MMU TLB exceptions are handled directly, as they must not
+ *   re-enable interrupts at any point before returning to the trapping
```
what you mean here ? syscall handler is expected to run with interruptions enabled
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/kernel/mem_protect/userspace/src/main.c#L934

andyross replied:
Maybe the typography is confusing?  Two items in a bulleted list:

* System calls have their own optimizer handler
* TLB exceptions don't unmask EXCM

Syscalls absolutely do unmask interrupts, obviously.

ceolin commented on arch/xtensa/core/xtensa-win0.S:
```diff
@@ -0,0 +1,523 @@
+	s32i a0, a10, ___thread_t_arch_OFFSET + __xtensa_win0_ctx_t_user_ps_OFFSET
+
+	/* Further cook PS to set RING=0 and UM=0, then set our sp and unmask */
+	movi a1, ~(PS_RING_MASK | PS_UM_MASK)
```
why are you clearing up `UM` ? As far as I could see kernel threads and user threads are using user vector mode, the exception is only here ...

In other hand, you have to unmask EXCM, since syscall handlers expects it.

andyross replied:
Just performance, really.  The user vector has more work to do than the kernel vector: it needs to check for syscall entry.  Obviously system call handlers won't be making syscalls of their own, so there's no point.

And the code already removed EXCM from this the PS value in a0 a few lines above.

andyross replied:
Oh, and actually in this version of the PR, all threads are initialized with a PS of zero, so all threads are kernel mode by default.  Shimming the userspace stuff on top as we speak, at which point we'll be consistent between RING and UM (i.e. you either have RING=3/UM=1 or RING=0/UM=0, and no other states)

ceolin replied:
Oh ok, just noticed that thread `ctx` `ps` is not set :P

ceolin commented on arch/xtensa/core/xtensa-win0.S:
```diff
@@ -0,0 +1,523 @@
+	l32i a1, a1, 0
+#endif
+	wsr a0, PS
+
```
assuming that you have to re-enable interruptions / exceptions before jump into the handler
you can be switched after setting `PS`  and  syscall parameters are not saved in this context stack ...

andyross replied:
The interrupted userspace a0/a1/pc/ps values (the ones we need to preserve) are already stashed in the thread struct at this point.  All a context switch arriving now has to do is save our kernel space GPRs.

ceolin replied:
that happens only with interrutptions ritght ? inside `must_switch`

I remember having seen case where a syscall triggers a context switch  without an interruption happens and AFAIU `xtensa_switch` does not save them.

ceolin commented:
once I built the toolchain, I had to add `+CONFIG_XTENSA_CALL0_ABI=y` in `boards/xtensa/qemu_xtensa/qemu_xtensa_mmu_defconfig` to build the sample. Trying to run the sample I get:
```
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU  ** FATAL EXCEPTION
 ** CPU qemu-system-xtensa: terminating on signal 2
ninja: build stopped: interrupted by user.
```
it seems to be happening in ` printk("Launching crashing thread %p\n", &boom);`

Removing some printks the code runs a little bit further but it stucks with:
```
Flagging medium priority interrupt                                       
swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint!
 (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nest
ed = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = s
wint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! 
(nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (ne
sted = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested =
 swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint
! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nes
ted = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = 
swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint!
 (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nested = swint! (nest
ed = swint! (nested = swint! (nested = swin
```

It seems I am missing something to reproduce the sample, just after that I noticed that the build it generates a lot of warnings:
```
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_floatsidf.o): warning: incompatible Xtensa configuration (ABI do
es not match)                                                                                                                                                                                                                                                                                          
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_floatdidf.o): warning: incompatible Xtensa configuration (ABI do
es not match)                                                                                                                                                                                                                                                                                          
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_truncdfsf2.o): warning: incompatible Xtensa configuration (ABI d
oes not match)                                                                                                                                                                                                                                                                                         
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_extendsfdf2.o): warning: incompatible Xtensa configuration (ABI 
does not match)                                                                                                                                                                                                                                                                                        
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_divdi3.o): warning: incompatible Xtensa configuration (ABI does 
not match)                                                               
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_moddi3.o): warning: incompatible Xtensa configuration (ABI does 
not match)                                                               
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_udivdi3.o): warning: incompatible Xtensa configuration (ABI does
 not match)                                                              
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: /home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/libgcc.a(_umoddi3.o): warning: incompatible Xtensa configuration (ABI does
 not match)                                                              
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: warning: zephyr/zephyr.elf has a LOAD segment with RWX permissions
/home/ceolin/x-tools/xtensa-dc233c_zephyr-elf/lib/gcc/xtensa-dc233c_zephyr-elf/13.2.0/../../../../xtensa-dc233c_zephyr-elf/bin/ld.bfd: warning: zephyr/zephyr.elf has a LOAD segment with RWX permissions
```

What is missing in the toolchain ?

andyross commented on arch/xtensa/core/mmu.c:
```diff
@@ -0,0 +1,278 @@
+		for(int i = 0; i < nent; i++) {
+			uint32_t arg = (i << 12) | way;
+
+			__asm__ volatile("idtlb %0; isync; iitlb %0; isync" :: "r"(arg));
```
Heh, indeed.  This is some debugging voodoo that got left in.

andyross commented:
@ceolin:
> It seems I am missing something to reproduce the sample, just after that I noticed that the build it generates a lot of warnings:

Yeah, that's what you get if you have a windowed toolchain (strictly the libgcc.a in the toolchain) building against a call0 Zephyr, or vice versa.  It will crash any time it hits the ENTRY instruction at the start of a libgcc instruction (because PS.WOE=0).  And indeed printk() is one of the big offenders, as it uses 64 bit math by default.  If you look way upthread, there are some kconfigs I found that suppress enough 64 bit-isms to make the test run, and that probably works still.

Another option is to use a cadence toolchain, as those are multilib and don't have this problem.

Finally remember @dcpleung had patches to the SDK working which build a multilib gcc; I never got to trying them myself but for sure that's a better solution than the hand-built one in my instructions above.  Maybe he could just hand you a copy?

ceolin commented:
> @ceolin:
> 
> > It seems I am missing something to reproduce the sample, just after that I noticed that the build it generates a lot of warnings:
> 
> Yeah, that's what you get if you have a windowed toolchain (strictly the libgcc.a in the toolchain) building against a call0 Zephyr, or vice versa. It will crash any time it hits the ENTRY instruction at the start of a libgcc instruction (because PS.WOE=0). And indeed printk() is one of the big offenders, as it uses 64 bit math by default. If you look way upthread, there are some kconfigs I found that suppress enough 64 bit-isms to make the test run, and that probably works still.

Yep, I know it. The point is that I follow the whole instructions to get sdk-ng and crosstool and to build them. I thought with that we had a proper libgcc ...

> 
> Another option is to use a cadence toolchain, as those are multilib and don't have this problem.
> 

One of the things you pointed in the other pr is not having a way to test in CI, that is where we are here ...

andyross commented:
> One of the things you pointed in the other pr is not having a way to test in CI, that is where we are here ...

I know.  :)  This is still a DNM PR for sure, pending a toolchain release.

I **swear** I'm not trying to monopolize this or turn it into a fight.   We need one codebase that meets all requirements (though to be fair this one has/will-have some genuine advantages in performance/size).  Clearly neither is there yet, the intent is to turn this one into a merged PR with the userspace support from the other.  But I'm very happy to go in the other direction too.

ceolin commented on arch/xtensa/core/mmu.c:
```diff
@@ -0,0 +1,278 @@
+	regs->ptevaddr = CONFIG_XTENSA_MMU_PTE_BASE + user_asid * 0x400000;
+
+	/* The ptables code doesn't add the mapping for the l1 page itself */
+	l1_page[l1_idx(regs->ptevaddr)] = (uint32_t) l1_page;
```
Manipulating the page table here does not look nice if the goal is to split the mmu / page table logic

ceolin replied:
the cache bits may not match as well ...

ceolin commented on arch/xtensa/core/mmu.c:
```diff
@@ -0,0 +1,278 @@
+	__ASSERT_NO_MSG(user_asid != 1 && user_asid < CONFIG_XTENSA_MMU_ASID_MAX);
+
+	/* We don't use ring 1/2, ring 0 ASID must be 1 */
+	regs->rasid = (user_asid << 24) | 0x000001;
```
As far as I know each ring in RASID must be different, we can't leave rings 1/2 with the same value
